Document mutability and metaclass interactions (#154)
antonagestam/phantom-types: docs/pages/composing-types.rst

.. code-block:: python

    class UTCDateTime(datetime.datetime, Phantom, predicate=is_utc):
        ...

Mutability
==========

Phantom types are completely incompatible with mutable data and should never be used to
narrow a mutable type. The reason is that there is no way for a type checker to detect
that a mutation changes an object to no longer satisfying the predicate of a phantom
type. For example:

.. code-block:: python

    # A phantom type that checks that a list has more than 2 items.
    class HasMany(list, Phantom, predicate=count(greater(2))):
        ...

    # The check will pass because the list *currently* has 3 items in it.
    instance = HasMany.parse([1, 2, 3])

    # But! Lists are mutable, so nothing is stopping us from removing an item, at this
    # point the list will only have 2 items and won't satisfy the predicate of the
    # HasMany type anymore.
    del instance[-1]

    # There is no way for a type checker to now that the predicate isn't fulfilled
    # anymore, so the revealed type here will still be HasMany.
    reveal_type(instance)  # Revealed type is HasMany

In some cases phantom-types tries to be smart and disallow using mutable types as

> bounds, but in the general case this isn't possible to detect and so relies on developer
> discipline to stay away from mutable data types.

Author: ```suggestion
bounds, but in the general case this isn't possible to detect and so it's up to you as a
developer to make sure to not mix mutable data with phantom types.
```

Metaclass conflicts
===================

Phantom types are implemented using a metaclass. When creating a phantom type that

> narrows on a type that also uses metaclasses it's common to stumble into a metaclass

Author: ```suggestion
narrows on a type that also uses a metaclass it's common to stumble into a metaclass
```

conflict. The usual solution to such situation is to create a new metaclass that

> inherits both existing metaclasses, and use it on the new type.

Author: ```suggestion
inherits both existing metaclasses and base the new type on it.
```
